core/src/main/java/software/amazon/awssdk/core/async/FileAsyncRequestBody.java

```diff
@@ -251,11 +248,30 @@ private void closeFile() {
                 inputChannel.close();
             } catch (IOException e) {
                 synchronized (subscriber) {
-                    subscriber.onError(e);
+                    signalOnError(e);
                 }
             }
         }
 
+        private void signalOnNext(ByteBuffer bb) {
+            if (!done.get()) {
```

Comment:
Is there still a race condition here if onComplete signals after we check done?

Reply:
Ah yeah, It could happen for `onError` though if we signal for the non positive demand case. `onComplete` and `onNext` are only signaled within the same thread, so I don't think that could happen for that case.

Update so all signaling to synchronizes on `this`